dataframe-csv/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/io/readDelim.kt

```diff
@@ -0,0 +1,345 @@
+import org.jetbrains.kotlinx.dataframe.documentation.DelimParams.IGNORE_EXCESS_COLUMNS
+import org.jetbrains.kotlinx.dataframe.documentation.DelimParams.IGNORE_SURROUNDING_SPACES
+import org.jetbrains.kotlinx.dataframe.documentation.DelimParams.INPUT_STREAM_READ
+import org.jetbrains.kotlinx.dataframe.documentation.DelimParams.PARSER_OPTIONS
+import org.jetbrains.kotlinx.dataframe.documentation.DelimParams.PARSE_PARALLEL
+import org.jetbrains.kotlinx.dataframe.documentation.DelimParams.QUOTE
+import org.jetbrains.kotlinx.dataframe.documentation.DelimParams.READ_LINES
+import org.jetbrains.kotlinx.dataframe.documentation.DelimParams.SKIP_LINES
+import org.jetbrains.kotlinx.dataframe.documentation.DelimParams.TRIM_INSIDE_QUOTED
+import org.jetbrains.kotlinx.dataframe.impl.ColumnNameGenerator
+import org.jetbrains.kotlinx.dataframe.io.ColType
+import org.jetbrains.kotlinx.dataframe.io.Compression
+import org.jetbrains.kotlinx.dataframe.io.DEFAULT_NULL_STRINGS
+import org.jetbrains.kotlinx.dataframe.io.skippingBomCharacters
+import org.jetbrains.kotlinx.dataframe.io.toKType
+import org.jetbrains.kotlinx.dataframe.io.useDecompressed
+import java.io.InputStream
+import java.math.BigDecimal
+import java.net.URL
+import kotlin.reflect.KType
+import kotlin.reflect.full.withNullability
+import kotlin.reflect.typeOf
+import kotlin.time.Duration
+
+/**
+ * Implementation to read delimiter-separated data from an [InputStream] based on the Deephaven CSV library.
+ *
+ * @include [INPUT_STREAM_READ]
+ * @param delimiter The field delimiter character. The default is ',' for CSV, '\t' for TSV.
+ * @include [HEADER]
+ * @include [COMPRESSION]
+ * @include [COL_TYPES]
+ * @include [SKIP_LINES]
+ * @include [READ_LINES]
+ * @include [PARSER_OPTIONS]
+ * @include [IGNORE_EMPTY_LINES]
+ * @include [ALLOW_MISSING_COLUMNS]
+ * @include [IGNORE_EXCESS_COLUMNS]
+ * @include [QUOTE]
+ * @include [IGNORE_SURROUNDING_SPACES]
+ * @include [TRIM_INSIDE_QUOTED]
+ * @include [PARSE_PARALLEL]
+ * @include [ADDITIONAL_CSV_SPECS]
+ */
+internal fun readDelimImpl(
+    inputStream: InputStream,
+    delimiter: Char,
+    header: List<String> = HEADER,
+    compression: Compression<*> = COMPRESSION,
+    colTypes: Map<String, ColType> = COL_TYPES,
+    skipLines: Long = SKIP_LINES,
+    readLines: Long? = READ_LINES,
+    parserOptions: ParserOptions = PARSER_OPTIONS,
+    ignoreEmptyLines: Boolean = IGNORE_EMPTY_LINES,
+    allowMissingColumns: Boolean = ALLOW_MISSING_COLUMNS,
+    ignoreExcessColumns: Boolean = IGNORE_EXCESS_COLUMNS,
+    quote: Char = QUOTE,
+    ignoreSurroundingSpaces: Boolean = IGNORE_SURROUNDING_SPACES,
+    trimInsideQuoted: Boolean = TRIM_INSIDE_QUOTED,
+    parseParallel: Boolean = PARSE_PARALLEL,
+    additionalCsvSpecs: CsvSpecs? = ADDITIONAL_CSV_SPECS,
+): DataFrame<*> {
+    // set up the csv specs
+    val csvSpecs = with(CsvSpecs.builder()) {
+        if (additionalCsvSpecs != null) from(additionalCsvSpecs)
```

Comment:
Parameters set in `additionalCsvSpecs` are now overridden by the other (default) parameters. Do we think this is intended behavior? Or should we let `additionalCsvSpecs` overwrite all parameters given here?

Additionally, we could make a separate public function which takes _just_ a `CsvSpecs` instance and skips all our default parameters and DF logic. WDYT?

Reply:
Fixed by making it a lambda running on the builder :)